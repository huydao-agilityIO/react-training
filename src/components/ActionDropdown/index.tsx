import { memo } from 'react';
import {
  IconButton,
  List,
  ListItem,
  Popover,
  PopoverBody,
  PopoverContent,
  PopoverTrigger,
  Stack,
  useDisclosure
} from '@chakra-ui/react';

// Svg
import { EllipsisVertical } from '@shared/SVG';

interface Action {
  label: string;
  onClick?: () => void;
}

interface ActionDropdownProps {
  actions: Action[];
}

const ActionDropdown = ({ actions }: ActionDropdownProps) => {
  const { isOpen, onOpen, onClose } = useDisclosure();

  return (
    <Popover isOpen={isOpen} onClose={onClose} placement="bottom-start">
      <PopoverTrigger>
        <IconButton
          aria-label="show-dropdown"
          variant="default"
          color="secondary.150"
          _hover={{
            color: 'secondary.600'
          }}
          icon={<EllipsisVertical color="secondary.150" />}
          onClick={onOpen}
        />
      </PopoverTrigger>
      <PopoverContent w={{ base: 20, md: 40 }} paddingTop={3} paddingBottom={5}>
        <PopoverBody>
          <Stack>
            <List>
              {actions?.map(({ label, onClick }) => (
                <ListItem
                  fontSize="xs"
                  fontWeight="light"
                  color="secondary.200"
                  px={5}
                  py={1}
                  _hover={{
                    cursor: 'pointer',
                    color: 'primary.300'
                  }}
                  key={label}
                  onClick={onClick}>
                  {label}
                </ListItem>
              ))}
            </List>
          </Stack>
        </PopoverBody>
      </PopoverContent>
    </Popover>
  );
};

export default memo(ActionDropdown);
