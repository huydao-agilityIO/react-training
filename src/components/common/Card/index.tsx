import { memo } from 'react';
import {
  Card,
  CardBody,
  CardHeader,
  Avatar,
  Heading,
  Text
} from '@chakra-ui/react';

interface CardComponentProps {
  name: string;
  description: string;
  url?: string;
}

const CardComponent = ({ url = '', name, description }: CardComponentProps) => {
  return (
    <Card
      direction="column"
      alignItems="center"
      boxShadow="sm"
      py={12}
      w={{ sm: '3xs', md: '2xs', lg: 'xs' }}>
      <Avatar src={url} name={name} size="xl" marginBottom={6} />
      <CardHeader p={0}>
        <Heading variant="secondary" size="md" marginBottom={1}>
          {name}
        </Heading>
      </CardHeader>
      <CardBody p={0}>
        <Text variant="helper" size="md" fontWeight="medium">
          {description}
        </Text>
      </CardBody>
    </Card>
  );
};

export default memo(CardComponent);
