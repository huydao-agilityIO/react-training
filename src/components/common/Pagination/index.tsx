import { Button, HStack } from '@chakra-ui/react';
import { memo } from 'react';

interface PaginationProps {
  currentPage: number;
  totalPage: number;
  onChangePage: (pageNumber: number) => void;
}

const Pagination = ({
  currentPage,
  totalPage,
  onChangePage
}: PaginationProps) => {
  const pageNumbers: number[] = Array(totalPage)
    .fill(null)
    .map((_, index) => index + 1);

  const handleNextPage = () => {
    onChangePage(currentPage + 1);
  };

  const handlePreviousPage = () => {
    onChangePage(currentPage - 1);
  };

  return (
    <HStack>
      <Button onClick={handlePreviousPage} isDisabled={currentPage === 1}>
        Previous
      </Button>{' '}
      {pageNumbers.map((number) => {
        const handleChangePage = () => {
          onChangePage(number);
        };
        return (
          <Button
            key={number}
            isDisabled={currentPage === number}
            onClick={handleChangePage}>
            {number}
          </Button>
        );
      })}
      <Button onClick={handleNextPage} isDisabled={currentPage === totalPage}>
        Next
      </Button>
    </HStack>
  );
};

export default memo(Pagination);
