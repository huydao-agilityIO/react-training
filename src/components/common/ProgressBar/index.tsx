import {
  Progress,
  Stack,
  ProgressProps,
  Text,
  Heading
} from '@chakra-ui/react';
import { memo, useCallback } from 'react';

interface ProgressBarComponentProps extends ProgressProps {
  variant?: 'default' | 'success' | 'warning' | 'danger';
  label?: string;
  value?: number;
}

const ProgressBarComponent = ({
  label,
  variant,
  value,
  size
}: ProgressBarComponentProps) => {
  const renderValueProgress = useCallback(() => {
    const MAPPING_COLOR_VALUE_PROGRESS = {
      default: { color: 'primary.500' },
      success: { color: 'success.200' },
      danger: { color: 'danger.200' },
      warning: { color: 'warning.200' }
    };

    return (
      <Text
        size="xs"
        position="absolute"
        top="-2%"
        display="block"
        width="full"
        textAlign="right"
        {...MAPPING_COLOR_VALUE_PROGRESS[variant ?? 'default']}>
        {value}%
      </Text>
    );
  }, [value, variant]);

  return (
    <Stack position="relative">
      <Heading
        textTransform="uppercase"
        color="dark.200"
        fontSize="xs"
        lineHeight="shorter">
        {label}
      </Heading>
      {renderValueProgress()}
      <Progress variant={variant} value={value} borderRadius="md" size={size} />
    </Stack>
  );
};

export default memo(ProgressBarComponent);
