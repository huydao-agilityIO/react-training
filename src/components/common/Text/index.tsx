import { memo } from 'react';
import { Link, Text, TextProps } from '@chakra-ui/react';
import { Link as LinkRouter } from 'react-router-dom';

interface TextLinkProps extends TextProps {
  link: string;
  pathRouter: string;
}

export const TextErrorMessage = memo(({ children, ...props }: TextProps) => (
  <Text {...props} variant="error" size="xs">
    {children}
  </Text>
));

export const TextHelper = memo(({ children, ...props }: TextProps) => (
  <Text {...props} variant="helper" size="md">
    {children}
  </Text>
));

export const TextLink = memo(
  ({ children, link, pathRouter, ...props }: TextLinkProps) => (
    <Text {...props} variant="link" size="lg" fontWeight="medium">
      {children}{' '}
      <Link
        color="secondary.400"
        _hover={{ textDecoration: 'none' }}
        as={LinkRouter}
        to={pathRouter}>
        {link}
      </Link>
    </Text>
  )
);
