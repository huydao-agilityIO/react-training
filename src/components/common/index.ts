export { default as Card } from './Card';
export { default as ProgressBar } from './ProgressBar';
export * from './Text';
