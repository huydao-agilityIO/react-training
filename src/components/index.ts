// Icons
export * from '../SVG';

// Common components
export * from './common';
