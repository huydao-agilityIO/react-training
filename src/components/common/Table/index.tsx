import { memo, useCallback } from 'react';
import {
  Spinner,
  Table,
  TableCaption,
  TableContainer,
  Tbody,
  Td,
  Thead,
  Tr
} from '@chakra-ui/react';

// Types
import { ContentMapping, HeadingMapping } from '@shared/types';

// Components
import { TextHelper } from '@shared/components';

interface DataTableProps {
  data: ContentMapping[][];
  headingMapping?: HeadingMapping[];
  isLoading?: boolean;
}

const DataTable = ({
  data,
  headingMapping = [],
  isLoading = false
}: DataTableProps) => {
  const renderBodyTable = useCallback(() => {
    if (isLoading) {
      return (
        <TableCaption paddingY={10}>
          <Spinner />
        </TableCaption>
      );
    }

    if (data.length === 0) {
      return (
        <TableCaption paddingY={10} fontFamily="body">
          <TextHelper>No Data Found</TextHelper>
        </TableCaption>
      );
    }

    return (
      <Tbody>
        {data.map((contentMapping: ContentMapping[], index) => (
          <Tr key={`HAD_${index}`}>
            {contentMapping.map(({ customRender, value, cellStyle }) => (
              <Td {...cellStyle}>{value ?? customRender?.()}</Td>
            ))}
          </Tr>
        ))}
      </Tbody>
    );
  }, [data, isLoading]);

  return (
    <TableContainer w={{ sm: 'full', xl: '6xl' }} bg="light.300">
      <Table>
        <Thead>
          <Tr bg="primary.300">
            {headingMapping?.map(({ key, label, cellStyle }) => (
              <Td key={key} {...cellStyle}>
                {label}
              </Td>
            ))}
          </Tr>
        </Thead>
        {renderBodyTable()}
      </Table>
    </TableContainer>
  );
};

export default memo(DataTable);
