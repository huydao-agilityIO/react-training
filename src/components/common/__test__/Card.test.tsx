// Component
import { Card } from '@shared/components/common';

const hospitalStaff = {
  url: 'https://imgs.search.brave.com/YbuHfY7YK6CuWXrBBzv3nQLty9mGR_hC7L7CsaLjJE8/rs:fit:860:0:0/g:ce/aHR0cHM6Ly9pLnBp/bmltZy5jb20vb3Jp/Z2luYWxzLzZiLzUy/LzlmLzZiNTI5ZmRl/MTYxNjU5MmYxMzVj/NGMxMjVjNDdlNGEx/LmpwZw',
  name: 'Lionel Messi',
  description: 'Doctor'
};

describe('Card Component', () => {
  it('renders correctly with props', () => {
    const { url, name, description } = hospitalStaff || {};

    const { asFragment } = testLibReactUtils.render(
      <Card url={url} name={name} description={description} />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
