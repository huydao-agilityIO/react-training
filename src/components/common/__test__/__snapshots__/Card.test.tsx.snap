// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component renders correctly with props 1`] = `
<DocumentFragment>
  <div
    class="chakra-card css-1nnsn8o"
  >
    <span
      class="chakra-avatar css-2yxd85"
    >
      <div
        aria-label="Lionel Messi"
        class="chakra-avatar__initials css-1ebyn6"
        role="img"
      >
        LM
      </div>
    </span>
    <div
      class="chakra-card__header css-old1by"
    >
      <h2
        class="chakra-heading css-1cj8ah8"
      >
        Lionel Messi
      </h2>
    </div>
    <div
      class="chakra-card__body css-1830czf"
    >
      <p
        class="chakra-text css-340eiw"
      >
        Doctor
      </p>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
